### PR TITLE
Transform things

### PR DESCRIPTION
**Pull request by lukewestby**
This is everything except `transform-origin`, still haven't thought all the way through overlaying `left | right | top | bottom` with lengths.

Thanks!

Ref #23

### CODE REVIEW

File: src/Css.elm [hunk 1]
```diff
@@ -833,6 +911,366 @@ n =
...
+{- TRANSFORMS -}
+
+
+cssFunction : String -> (a -> String) -> List a -> String
+cssFunction funcName converter args =
+    args
+        |> List.map converter
+        |> String.join ", "
+        |> (++) (funcName ++ "(")
+        |> (flip (++)) ")"
```
Comment (rtfeldman) [inline src/Css.elm]: Cool! We should switch over `rgb` and `rgba` to use this

File: src/Css.elm [hunk 2]
```diff
@@ -833,6 +911,366 @@ n =
...
+
+
+{-| The [`matrix()`](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function#matrix()) transform-function.
+
+    ~ transform (matrix a b c d tx ty)
```
Comment (rtfeldman) [inline src/Css.elm]: It's best to give examples that suggest what the units are, e.g.

```
~ transform (matrix 0 0.5 1.5 2 3.5 4.5)
```

File: src/Css.elm (hunk 1 above)
Comment (lukewestby) [inline src/Css.elm]: Definitely! I initially planned to do `cssFunction`, `cssFunction2`, etc. like you had done with `prop1`, `prop2`, etc. but then I saw that `matrix()` accepts 16 arguments and css functions almost always accept the same data type for all arguments so I figured lifting to a `List` would be cool.

File: src/Css.elm (hunk 2 above)
Comment (lukewestby) [inline src/Css.elm]: Right on, will do from now on.

File: src/Css.elm [hunk 3]
```diff
@@ -833,6 +911,366 @@ n =
...
+-}
+matrix : number -> number -> number -> number -> number -> number -> Transform
+matrix a b c d tx ty =
+    cssFunction "matrix" numberToString [ a, b, c, d, tx, ty ] |> Transform
+
+
+{-| The [`matrix3d()`](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function#matrix3d()) transform-function.
+
+    ~ transform (matrix3d a1 a2 a3 a4 b1 b2 b3 b4 c1 c2 c3 c4 d1 d2 d3 d4)
```
Comment (rtfeldman) [inline src/Css.elm]: same suggestion here as https://github.com/rtfeldman/elm-css/pull/46/files#r48451304

File: src/Css.elm [hunk 4]
```diff
@@ -833,6 +911,366 @@ n =
...
+-}
+translate3d : Length -> Length -> Length -> Transform
+translate3d tx ty tz =
+    cssFunction "translate3d" lengthToString [ tx, ty, tz ] |> Transform
+
+
+{-| Sets [`transform`](https://developer.mozilla.org/en-US/docs/Web/CSS/transform)
+with a series of transform-functions. If an empty list is provided, the CSS
+output will be `none`, as if to state directly that the set of transforms
+applied to the current selector is the empty:
```
Comment (rtfeldman) [inline src/Css.elm]: I assume typo here - intended to be "the empty set" or something similar?

File: src/Css.elm [hunk 5]
```diff
@@ -833,6 +911,366 @@ n =
...
+used instead for convenience. `transform` expects exactly one transform-function
+and cannot be passed `none`:
+
+    ~ transform (matrix 1 2 3 4 5 6) -- transform: matrix(1, 2, 3, 4, 5, 6);
+
+If a collection of transforms is needed, use the `transforms` function with a
+populated list:
+
+    ~ transforms [ perspective 1, scale2 1 1.4 ]
+-}
```
Comment (rtfeldman) [inline src/Css.elm]: Love the detailed documentation! :heart_eyes:

File: src/Css.elm (hunk 4 above)
Comment (lukewestby) [inline src/Css.elm]: Should have been "is empty"

File: test/FixturesCss.elm [hunk 6]
```diff
@@ -186,3 +186,43 @@ manualUnderlineOnHoverStyle =
...
+                ]
+            ~ transforms
+                [ translate (px 1)
+                , translate2 (px 1) (px 1)
+                , translateX (px 1)
+                , translateY (px 1)
+                , translate3d (px 1) (px 1) (px 1)
+                ]
+            ~ transformBox viewBox
+            ~ transformStyle preserve3d
```
Comment (rtfeldman) [inline test/FixturesCss.elm]: Sweet tests! :100:

Comment (lukewestby) [inline test/FixturesCss.elm]: elm-test is super easy to use once it's all set up and runnable :smile:

**Comment by rtfeldman** [comment]
@lukewestby Looks good! I figure you probably want to make those couple of chanages before I merge this...ping me when ready to merge?

File: test/FixturesCss.elm (hunk 6 above)
Comment (rtfeldman) [inline test/FixturesCss.elm]: Yeah! :smile_cat:

**Comment by lukewestby** [comment]
@rtfeldman all set!

**Comment by rtfeldman** [comment]
Looks great! Will merge as soon as tests pass.

Thanks so much for this!

**Comment by lukewestby** [comment]
No problem! Actually I want to do an interactive rebase to fix that botched commit message real quick, i'll post one more comment when ready.

**Comment by lukewestby** [comment]
Okay, I think that worked. If you don't get to it first I can start looking into `transform-origin` on Sunday.

**Comment by rtfeldman** [comment]
Merged! :tada:

**Pull request merged**